File: Algorithms/Graphs/dfs.py
from collections import defaultdict

#lines = input()
graph = {}
visited = defaultdict(int)
#completed = []


def add_edge(from_vertex, to_vertex):
    graph.setdefault(from_vertex, []).append(to_vertex)
    graph.setdefault(to_vertex, [])


# while lines > 0:
#     edge = [int(c) for c in raw_input().split(" ")]
#     add_edge(edge[0], edge[1])
#     lines -= 1

# for test input
add_edge(1, 2)
add_edge(2, 3)


def dfs(completed, v):
    if v in completed:
        print "cycle found"

    if not visited[v]:
        completed.append(v)
        visited[v] = 1
        for av in graph[v]:
            dfs(completed, av)
            print "Dfs next line."

    print "end of function."

    # #completed.append(v)
    # if not visited[v]:
    #     completed.append(v)
    #     for av in graph[v]:
    #         #if not visited[av]:
    #         visited[av] = 1
    #         dfs(completed, av)


start_vertex = graph.keys()[0]
dfs([], start_vertex)

File: Algorithms/Graphs/cpp/dfs.cpp
// C++ DFS Graph Traversal
#include <iostream>
#include <vector>

using namespace std;

vector<int> adj[5];
bool visited[5];

void dfs(int node) {
    if (visited[node]) return;
    visited[node] = true;
    cout << node << endl;
    for(auto u: adj[node]) {
        dfs(u);
    }

}

int main() {
    adj[1].push_back(2);
    adj[1].push_back(4);
    adj[2].push_back(3);
    adj[2].push_back(5);
    adj[3].push_back(5);
    dfs(1);
}
